tests/integration_test.bzl

```diff
@@ -0,0 +1,63 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helpers for running integration test."""
+
+load("@bazel_binaries//:defs.bzl", "bazel_binaries")
+load(
+    "@rules_bazel_integration_test//bazel_integration_test:defs.bzl",
+    "bazel_integration_tests",
+    "integration_test_utils",
+)
+
+def rules_python_integration_test(name, workspace_path = None, bzlmod = False, tags = None, **kwargs):
+    workspace_path = workspace_path or name.removesuffix("_example")
+    test_runner = "//tests:simple_test_runner" if bzlmod else "//tests:legacy_test_runner"
+
+    bazel_integration_tests(
+        name = name,
+        workspace_path = workspace_path,
+        test_runner = test_runner,
+        bazel_versions = bazel_binaries.versions.all,
+        workspace_files = integration_test_utils.glob_workspace_files(workspace_path) + [
+            "//:distribution",
+        ],
+        tags = (tags or []) + [
+            # Upstream normally runs the tests with the `exclusive` tag. That
+            # tag also implies `local` by default. We don't really need the
+            # exclusion feature, but we do need the test to break the sandbox.
```

Comment:
Why do we need it to break the sandbox?

Reply:
If I don't have it break the sandbox, then the tests will fail:

Reply:
```
INFO: Repository bazel_features~1.1.1 instantiated at:
  <builtin>: in <toplevel>
Repository rule http_archive defined at:
  /tmp/bazel-working-directory/rules_python/_tmp/2d3bceb7df44f5c60fb62c99b332cef6/_bazel_phil/99580b3b5ce42da1b00cf311bc78b2b8/external/bazel_tools/tools/build_defs/repo/http.bzl:384:31: in <toplevel>
ERROR: An error occurred during the fetch of repository 'bazel_features~1.1.1':
   Traceback (most recent call last):
        File "/tmp/bazel-working-directory/rules_python/_tmp/2d3bceb7df44f5c60fb62c99b332cef6/_bazel_phil/99580b3b5ce42da1b00cf311bc78b2b8/external/bazel_tools/tools/build_defs/repo/http.bzl", line 145, column 45, in _http_archive_impl
                download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: error copying file: couldn't delete destination: /home/users/phil/.cache/bazel_repository/content_addressable/sha256/62c26e427e5cbc751024446927622e398a9dcdf32c64325238815709d11c11a8/tmp-44772af6-b733-44fd-b1a2-ea49aa2d0bcf (Read-only file system)
ERROR: <builtin>: fetching http_archive rule //:bazel_features~1.1.1: Traceback (most recent call last):
        File "/tmp/bazel-working-directory/rules_python/_tmp/2d3bceb7df44f5c60fb62c99b332cef6/_bazel_phil/99580b3b5ce42da1b00cf311bc78b2b8/external/bazel_tools/tools/build_defs/repo/http.bzl", line 145, column 45, in _http_archive_impl
                download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: error copying file: couldn't delete destination: /home/users/phil/.cache/bazel_repository/content_addressable/sha256/62c26e427e5cbc751024446927622e398a9dcdf32c64325238815709d11c11a8/tmp-44772af6-b733-44fd-b1a2-ea49aa2d0bcf (Read-only file system)
```
It uses the user's home directory by default. Not sure if there's a way to fix that. But that's outside the scope of this PR I think.

Reply:
I get the impression the inner bazel is trying to download bazel_features? Strictly speaking, it shouldn't need to because the outer bazel would have (or could have) already done that, but the inner bazel probably isn't aware of where the outer bazel downloaded things to. In any case, yeah, I agree its out of scope for this PR.

Bazel's own CI does something to optimize this, so we can ask them how they make it work (I recall something about they have a special target to eagerly fetch things, then somehow wire the inner bazel to read from there).